File: Sources/MDifyCore/Services/NativeOCRRoutingClient.swift
import Foundation

public enum NativeOCRRoutingError: LocalizedError, Equatable {
    case nativeOCRReturnedNoText

    public var errorDescription: String? {
        switch self {
        case .nativeOCRReturnedNoText:
            "Apple Vision OCR did not recognize any text. Try MDify OCR for RapidOCR fallback."
        }
    }
}

public struct NativeOCRRoutingClient: WorkerConverting {
    private static let imageExtensions: Set<String> = ["jpg", "jpeg", "png", "bmp", "tif", "tiff", "webp"]
    private static let minimumTextCharacters = 12
    private static let minimumAverageConfidence = 0.45

    public let workerKind: WorkerKind
    private let workerClient: any WorkerConverting
    private let rapidOCRWorkerClient: (any WorkerConverting)?
    private let nativeOCR: any NativeOCRRecognizing

    public init(
        workerKind: WorkerKind,
        workerClient: any WorkerConverting,
        rapidOCRWorkerClient: (any WorkerConverting)? = nil,
        nativeOCR: any NativeOCRRecognizing = VisionOCRService()
    ) {
        self.workerKind = workerKind
        self.workerClient = workerClient
        self.rapidOCRWorkerClient = rapidOCRWorkerClient
        self.nativeOCR = nativeOCR
    }

    public func convert(inputURL: URL, outputURL: URL) async throws -> WorkerResponse {
        if Self.isImage(inputURL) {
            return try await convertWithNativeOCR(inputURL: inputURL, outputURL: outputURL)
        }

        if Self.isPDF(inputURL) {
            return try await convertPDF(inputURL: inputURL, outputURL: outputURL)
        }

        return try await workerClient.convert(inputURL: inputURL, outputURL: outputURL)
    }

    private func convertPDF(inputURL: URL, outputURL: URL) async throws -> WorkerResponse {
        let preflightResponse = try await workerClient.convert(inputURL: inputURL, outputURL: outputURL)
        guard preflightResponse.ok else { return preflightResponse }

        let markdown = (try? String(contentsOf: outputURL, encoding: .utf8)) ?? ""
        guard Self.isAlmostEmptyText(markdown) else {
            return preflightResponse
        }

        return try await convertWithNativeOCR(inputURL: inputURL, outputURL: outputURL)
    }

    private func convertWithNativeOCR(inputURL: URL, outputURL: URL) async throws -> WorkerResponse {
        let result = try await nativeOCR.recognize(inputURL: inputURL)
        let quality = Self.quality(for: result.markdown)
        let isWeak = quality.isEmpty
            || quality.alphanumericCount < Self.minimumTextCharacters
            || result.averageConfidence < Self.minimumAverageConfidence

        if workerKind == .ocr, isWeak {
            return try await fallbackToRapidOCR(inputURL: inputURL, outputURL: outputURL)
        }

        guard !quality.isEmpty else {
            throw NativeOCRRoutingError.nativeOCRReturnedNoText
        }

        try writeNativeMarkdown(result.markdown, to: outputURL)
        var warnings: [String] = []
        if isWeak {
            warnings.append("Apple Vision OCR confidence was low; saved the native result without RapidOCR fallback.")
        }
        return WorkerResponse(
            ok: true,
            outputPath: outputURL.path,
            inputPath: inputURL.path,
            worker: "native",
            engine: "apple-vision",
            ocrUsed: true,
            warnings: warnings,
            errorCode: nil,
            message: nil
        )
    }

    private func fallbackToRapidOCR(inputURL: URL, outputURL: URL) async throws -> WorkerResponse {
        let fallbackClient = rapidOCRWorkerClient ?? workerClient
        let response = try await fallbackClient.convert(inputURL: inputURL, outputURL: outputURL)
        return response.appendingWarning("Apple Vision OCR was weak; used RapidOCR fallback.")
    }

    private func writeNativeMarkdown(_ markdown: String, to outputURL: URL) throws {
        try FileManager.default.createDirectory(
            at: outputURL.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        try markdown.write(to: outputURL, atomically: true, encoding: .utf8)
    }

    private static func isImage(_ url: URL) -> Bool {
        imageExtensions.contains(url.pathExtension.lowercased())
    }

    private static func isPDF(_ url: URL) -> Bool {
        url.pathExtension.lowercased() == "pdf"
    }

    private static func isAlmostEmptyText(_ markdown: String) -> Bool {
        quality(for: markdown).alphanumericCount < minimumTextCharacters
    }

    private static func quality(for markdown: String) -> TextQuality {
        let alphanumericCount = markdown.unicodeScalars.filter {
            CharacterSet.alphanumerics.contains($0)
        }.count
        return TextQuality(
            isEmpty: markdown.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
            alphanumericCount: alphanumericCount
        )
    }
}

private struct TextQuality {
    let isEmpty: Bool
    let alphanumericCount: Int
}

private extension WorkerResponse {
    func appendingWarning(_ warning: String) -> WorkerResponse {
        WorkerResponse(
            ok: ok,
            outputPath: outputPath,
            inputPath: inputPath,
            worker: worker,
            engine: engine,
            ocrUsed: ocrUsed,
            warnings: warnings + [warning],
            errorCode: errorCode,
            message: message
        )
    }
}

File: windows/MDify.Windows/Services/NativeOcrRoutingClient.cs
using System.IO;
using System.Text.RegularExpressions;
using MDify.Windows.Models;

namespace MDify.Windows.Services;

public sealed class NativeOcrRoutingException : Exception
{
    public NativeOcrRoutingException(string message, Exception? innerException = null)
        : base(message, innerException)
    {
    }
}

public sealed class NativeOcrRoutingClient : IWorkerConverting
{
    private static readonly IReadOnlySet<string> ImageExtensions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
    {
        ".jpg",
        ".jpeg",
        ".png",
        ".bmp",
        ".tif",
        ".tiff",
        ".webp"
    };

    private const int MinimumTextCharacters = 12;
    private const double MinimumAverageConfidence = 0.45;

    private readonly IWorkerConverting _workerClient;
    private readonly IWorkerConverting? _rapidOcrWorkerClient;
    private readonly INativeOcrService _nativeOcr;

    public NativeOcrRoutingClient(
        WorkerKind workerKind,
        IWorkerConverting workerClient,
        IWorkerConverting? rapidOcrWorkerClient = null,
        INativeOcrService? nativeOcr = null)
    {
        WorkerKind = workerKind;
        _workerClient = workerClient;
        _rapidOcrWorkerClient = rapidOcrWorkerClient;
        _nativeOcr = nativeOcr ?? new WindowsNativeOcrService();
    }

    public WorkerKind WorkerKind { get; }

    public async Task<WorkerResponse> ConvertAsync(
        string inputPath,
        string outputPath,
        CancellationToken cancellationToken)
    {
        if (IsImage(inputPath))
        {
            return await ConvertWithNativeOcrAsync(inputPath, outputPath, cancellationToken)
                .ConfigureAwait(false);
        }

        if (IsPdf(inputPath))
        {
            return await ConvertPdfAsync(inputPath, outputPath, cancellationToken)
                .ConfigureAwait(false);
        }

        return await _workerClient.ConvertAsync(inputPath, outputPath, cancellationToken)
            .ConfigureAwait(false);
    }

    private async Task<WorkerResponse> ConvertPdfAsync(
        string inputPath,
        string outputPath,
        CancellationToken cancellationToken)
    {
        var preflightResponse = await _workerClient.ConvertAsync(inputPath, outputPath, cancellationToken)
            .ConfigureAwait(false);
        if (!preflightResponse.Ok)
        {
            return preflightResponse;
        }

        var markdown = File.Exists(outputPath) ? await File.ReadAllTextAsync(outputPath, cancellationToken) : "";
        if (!IsAlmostEmptyText(markdown))
        {
            return preflightResponse;
        }

        return await ConvertWithNativeOcrAsync(inputPath, outputPath, cancellationToken)
            .ConfigureAwait(false);
    }

    private async Task<WorkerResponse> ConvertWithNativeOcrAsync(
        string inputPath,
        string outputPath,
        CancellationToken cancellationToken)
    {
        NativeOcrResult result;
        try
        {
            result = await _nativeOcr.RecognizeAsync(inputPath, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception error) when (error is not OperationCanceledException)
        {
            if (HasRapidOcrFallback)
            {
                return await FallbackToRapidOcrAsync(
                    inputPath,
                    outputPath,
                    "Windows Text Recognizer failed; used RapidOCR fallback.",
                    cancellationToken).ConfigureAwait(false);
            }

            throw new NativeOcrRoutingException(
                "Windows Text Recognizer failed and no RapidOCR fallback worker is bundled.",
                error);
        }

        var quality = MeasureQuality(result.Markdown);
        var isWeak = quality.IsEmpty
            || quality.AlphanumericCount < MinimumTextCharacters
            || result.AverageConfidence < MinimumAverageConfidence;

        if (isWeak && WorkerKind == WorkerKind.Ocr && !HasRapidOcrFallback)
        {
            throw new NativeOcrRoutingException(
                "Windows Text Recognizer was weak and no RapidOCR fallback worker is bundled.");
        }

        if (isWeak && HasRapidOcrFallback)
        {
            return await FallbackToRapidOcrAsync(
                inputPath,
                outputPath,
                "Windows Text Recognizer was weak; used RapidOCR fallback.",
                cancellationToken).ConfigureAwait(false);
        }

        if (quality.IsEmpty)
        {
            throw new NativeOcrRoutingException("Windows Text Recognizer did not recognize any text.");
        }

        Directory.CreateDirectory(Path.GetDirectoryName(outputPath)!);
        await File.WriteAllTextAsync(outputPath, result.Markdown, cancellationToken).ConfigureAwait(false);

        var warnings = new List<string>();
        if (isWeak)
        {
            warnings.Add("Windows Text Recognizer confidence was low; saved native result without RapidOCR fallback.");
        }

        return new WorkerResponse(
            true,
            outputPath,
            inputPath,
            "native",
            "windows-text-recognizer",
            true,
            warnings,
            null,
            null);
    }

    private bool HasRapidOcrFallback => WorkerKind == WorkerKind.Ocr && _rapidOcrWorkerClient is not null;

    private async Task<WorkerResponse> FallbackToRapidOcrAsync(
        string inputPath,
        string outputPath,
        string warning,
        CancellationToken cancellationToken)
    {
        if (_rapidOcrWorkerClient is null)
        {
            throw new NativeOcrRoutingException("RapidOCR fallback worker is not bundled.");
        }

        var response = await _rapidOcrWorkerClient.ConvertAsync(inputPath, outputPath, cancellationToken)
            .ConfigureAwait(false);
        return response with { Warnings = response.Warnings.Concat(new[] { warning }).ToArray() };
    }

    private static bool IsImage(string path)
    {
        return ImageExtensions.Contains(Path.GetExtension(path));
    }

    private static bool IsPdf(string path)
    {
        return string.Equals(Path.GetExtension(path), ".pdf", StringComparison.OrdinalIgnoreCase);
    }

    private static bool IsAlmostEmptyText(string markdown)
    {
        return MeasureQuality(markdown).AlphanumericCount < MinimumTextCharacters;
    }

    private static TextQuality MeasureQuality(string markdown)
    {
        var alphanumericCount = Regex.Matches(markdown, @"[\p{L}\p{Nd}]").Count;
        return new TextQuality(
            string.IsNullOrWhiteSpace(markdown),
            alphanumericCount);
    }

    private sealed record TextQuality(
        bool IsEmpty,
        int AlphanumericCount);
}
